proposals/data-objects.md

```diff
@@ -0,0 +1,76 @@
+# Data objects
+
+* **Type**: Design proposal
+* **Authors**: Alexander Udalov, Roman Elizarov, Pavel Mikhailovskii
+* **Status**: Experimental since 1.7.20, stable since 1.8.0
+
+
+## Introduction
+
+When using algebraic data types, it’s annoying to declare singleton values because there’s no way to get automatic toString similar to data classes. E.g. on JVM:
+```kotlin
+sealed class List<T> {
+  data class Cons<T>(val value: T) : List<T>()
+  object Nil : List<Nothing>()
+}
+
+fun main() {
+println(List.Cons(42))   // Cons(value=42)
+println(List.Nil)        // test.List$Nil@1d251891
+}
+```
+
+A workaround for the user is either to declare `toString `manually, which leads to boilerplate, or use data class with a placeholder default argument, which seems bizarre.
+
+## Other aspects
+
+* Default toString implementation differs on JVM, JS, Native
+    * In JS, toString returns [object Object]
+    * Native behaves similarly to JVM, but inner classes are delimited with ., i.e. test.List*.*Nil@1d251891 in the example above
+* The @hash part is meaningless even for non-ADT objects, because object is a singleton
+* Can’t declare an empty @JvmRecord class in Kotlin (KT-48155 (https://youtrack.jetbrains.com/issue/KT-48155/JvmRecord-does-not-allow-one-to-define-no-parameters))
+    * Probably not a problem
+
+### toString for all objects
+
+One possible solution is to generate toString for all objects unconditionally, which returns the simple (unqualified) name of the object.
+
+The advantage is that it’s more uniform, and the least surprising for the users.
+
+There are some issues though:
+
+* It’s likely a major breaking change
+* It leads to extra bytecode which is often unnecessary
+* Simple name instead of FQ name might be confusing, especially for cases like companion object
+
+### Data object
+
+Another possible solution is to allow the data modifier on objects, which will lead to toString being generated as for data classes.
+
+Other data class methods should not be generated:
+
+* equals and hashCode work fine already since it’s a singleton
```

Comment:
This part conflicts with the subsequent proposal without giving any rationale.

Reply:
> This part conflicts with the subsequent proposal without giving any rationale.

Thanks for reviewing. Please take another look.